objectbox/lib/src/transaction.dart

```diff
@@ -0,0 +1,84 @@
+import 'dart:ffi';
+
+import 'bindings/bindings.dart';
+import 'bindings/helpers.dart';
+import 'modelinfo/entity_definition.dart';
+import 'store.dart';
+
+enum TxMode {
+  Read,
+  Write,
+}
+
+// TODO enable annotation once meta:1.3.0 is out
+// @internal
+class Transaction {
+  final Store _store;
+  final bool _isWrite;
+  final Pointer<OBX_txn> _cTxn;
+  final _cursors = <int, CursorHelper>{};
+  bool _closed = false;
+
+  Pointer<OBX_txn> get ptr => _cTxn;
+
+  Transaction(this._store, TxMode mode)
+      : _isWrite = mode == TxMode.Write,
+        _cTxn = mode == TxMode.Write
+            ? C.txn_write(_store.ptr)
+            : C.txn_read(_store.ptr) {
+    checkObxPtr(_cTxn, 'failed to create transaction');
+  }
+
+  void _finish(bool successful) {
+    if (_isWrite) {
+      try {
+        markSuccessful(successful);
+      } finally {
+        close();
+      }
+    } else {
+      close();
+    }
+  }
+
+  void commitAndClose() => _finish(true);
+
+  void abortAndClose() => _finish(false);
+
+  void markSuccessful([bool successful = true]) =>
+      checkObx(C.txn_mark_success(_cTxn, successful));
+
+  void close() {
+    if (_closed) return;
+    _closed = true;
+    _cursors.values.forEach((c) => c.close());
+    _cursors.clear();
+    checkObx(C.txn_close(_cTxn));
+  }
+
+  /// Returns a cursor for the given entity. No need to close it manually.
+  /// Note: the cursor may have already been used, don't rely on its state!
+  CursorHelper cursor(EntityDefinition entity) {
+    final entityId = entity.model.id.id;
+    if (_cursors.containsKey(entityId)) {
+      return _cursors[entityId];
+    }
+    return _cursors[entityId] = CursorHelper(_store, _cTxn, entity, _isWrite);
+  }
+
+  /// Executes a given function inside a transaction.
+  ///
+  /// Returns type of [fn] if [return] is called in [fn].
+  static R execute<R>(Store store, TxMode mode, R Function() fn) {
+    final tx = Transaction(store, mode);
+    try {
+      if (tx._isWrite) tx.markSuccessful(true);
+      return fn();
```

Comment:
I know this is the same behavior as before, but shouldn't the transaction be marked successful *after* `fn` executed? E.g. similar to `put()`?

Reply:
Yes, in the perfect world where we wouldn't have to define a variable storing the result of `fn()` before returning it after a `markSuccessful(true)` call. This is more efficient, assuming transactions "usually" don't fail

Reply:
So this is intended, maybe add a note?

Reply:
OK